src/main/java/backend/RepoIO.java

```diff
@@ -51,14 +52,32 @@ public RepoIO(boolean isTestMode, boolean enableTestJSON) {
 
     public CompletableFuture<Model> openRepository(String repoId) {
         if (repoStore.isRepoStored(repoId)) {
-            return repoStore.loadRepository(repoId)
-                .thenCompose(this::updateModel)
-                .exceptionally(withResult(new Model(repoId)));
+            return loadRepositoryFromStore(repoId);
         } else {
-            return repoSource.downloadRepository(repoId)
+            return downloadRepositoryFromSource(repoId);
+        }
+    }
+
+    private CompletableFuture<Model> loadRepositoryFromStore(String repoId) {
+        return repoStore.loadRepository(repoId)
+                .thenCompose(this::updateModel)
+                .exceptionally(ex -> {
+                    try {
```

Comment:
There *are* lots of different factors at play here. We should pick something that achieves the right tradeoffs.

Personally I'm for either `withFuture` (where error-handling is left implicit) or a method like `downloadRepoNow` (which factors out the blocking `get`, keeping error-handling explicit). Either one keeps things at the same level of abstraction *within the method*.

While there may be some asymmetry in how `load` and `download` work *as methods*, this is already the highest level of abstraction in which they exist (`RepoIO` is supposed to abstract away differences between GitHub and cache), so I'm not sure we can do anything about it. Anything higher up is meant to be insulated from caching details.

Blocking and updating the model in `exceptionally` should be fine. Anything downstream in the chain of operations will receive the same data, just a bit later. In that sense there are no side effects.

Handling exceptions in a generic manner has tradeoffs, as you mentioned. There are more flexible alternatives to `exceptionally` which give you the Throwable to work with, though I'm not sure that much explicitness is necessary. Maybe we could move the `exceptionally` clause higher, to make it explicit that we're not handling the failure of `updateModel`?

```java
return repoStore.loadRepository(repoId)
    .exceptionally(/* download repo */);
    .thenCompose(this::updateModel)
    .exceptionally(withResult(new Model(repoId)));
```

Speaking of which, would updating the empty model handle the download, i.e. would this work?

```java
return repoStore.loadRepository(repoId)
    .exceptionally(withResult(new Model(repoId)));
    .thenCompose(this::updateModel)
    .exceptionally(withResult(new Model(repoId)));
```